packages/elements/src/tree/elements/tree.ts

```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
    * @override
    */
   protected update (changedProperties: PropertyValues): void {
+    super.update(changedProperties);
     if (changedProperties.has('noRelation') || changedProperties.has('multiple')) {
       this.manager.setMode(this.mode);
     }
-    super.update(changedProperties);
+
+    if (changedProperties.has('query') || changedProperties.has('data')) {
+      this.filterItems();
+    }
+  }
+
+  /**
+   * Filter the internal items by query changes items' hidden state
+   * @returns {void}
+   */
+  protected filterItems (): void {
+    // if filter is null, it is off and external app is responsible
+    if (this.filter) {
+      const filter = this.filter;
+      const items = this.queryItems((item): boolean => {
+        // Do not filter hidden items
+        if (item.hidden) {
+          return false;
+        }
+
+        const result = filter(item);
+        if (result) {
+          this.manager.includeItem(item);
+        }
+        else {
+          this.manager.excludeItem(item);
+        }
+
+        return result;
+      }).slice();
+
+      // Do not expand EMS if there is no filter applied
+      if (this.query) {
+        /**
+         * Include all descendants of item list to make them all
+         * are accessible which user can navigate into the nested data
+         */
+        this.addItemDescendantsToRender(items);
+        this.addExpandedAncestorsToRender(items);
+      }
+    }
+  }
+
+  /**
+   * Utility method
+   * Adds descendants for each item passed
+   * @param items List of child items
+   * @returns {void}
+   */
+  public addItemDescendantsToRender (items: T[]): void {
+    items.forEach((item) => {
+      /**
+       * Prevent tree keep the expand state of item which does not matched with query,
+       * collapse them to make the result of filtering is crystal clear
+       */
+      if (this.manager.isItemExpanded(item)) {
+        this.manager.collapseItem(item);
+      }
+
+
+      /**
+       * show all descendants of items to make them all are selectable
+       * and user can navigate into nested data
+       */
+      const children = this.composer.getItemChildren(item);
+      if (children.length) {
+        this.addNestedItemsToRender(children, items);
+      }
+    });
+  }
+
+  /**
+   * Utility method
+   * Add nested children of item list
+   * @param items List of items
+   * @param excludeItems List of exclude items
+   * @returns {void}
+   */
+  protected addNestedItemsToRender (items: readonly T[], excludeItems: readonly T[]): void {
+    items.forEach(item => {
+      // Skip excluding item
+      if (!excludeItems.includes(item)) {
+        // Add item and nested children
+        if (this.composer.getItemPropertyValue(item, 'hidden') === true) {
```

Comment:
Why do you check the hidden state here? Just include the item.

Reply:
Why is that?

Reply:
OK, checking the hidden is not need anymore as you said because of the fixing code in the above conversation. I'll remove it.

Reply:
Removed